---
title: "Burrows Bay Octopus Phylogeny, edited for Cheyne's Thesis"
author: "Kirt Onthank"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_depth: '3'
    df_print: paged
  pdf_document:
    toc: yes
    number_sections: yes
    toc_depth: 3
---

# Loading Libraries
```{r Libraries, message=FALSE, warning=FALSE}
library(ape)
library(xlsx)
library(insect)
library(aphid)
library(DECIPHER)
library(magrittr)
library(phangorn)
library(rwty)
library(stringi)
library(Biostrings)
library(kableExtra)
library(knitr)
library(phytools)
```


This function to convert DNAbin objects used by the ape package to DNAStringSet objects used by the DECIPHER package was written by Joel Nitta, available on his github (https://gist.github.com/joelnitta/6f30a7c0f1c83d78c76a5469e935d56f)
```{r}
# Convert ape::DNAbin format to Biostrings::DNAStringSet format,
# optionally removing gaps
DNAbin_to_DNAstringset <- function (seqs, remove_gaps = TRUE) {
  if(isTRUE(remove_gaps)) {
  seqs %>% as.list() %>% as.character %>% 
      lapply(.,paste0,collapse="") %>% 
      lapply( function (x) gsub("-", "", x)) %>% 
      unlist %>% Biostrings::DNAStringSet()
  } else {
    seqs %>% as.list() %>% as.character %>% 
      lapply(.,paste0,collapse="") %>% 
      unlist %>% Biostrings::DNAStringSet()
  }
}
```


# Preparing the dataset
## Reading in accession numbers
First I read in the accession numbers and 
```{r}
access=read.csv("Octopus burrowing - data.csv",na.strings = "")
colnames(access)=c("Species","burrow","keel","reference","COI","COIII","x12s","x16s","x28s","Rhodopsin","Pax-6","cytb","notes")
access=access[,1:13] #Dropping the junk columns
access=access[,-c(4,13)] #Dropping notes and references columns
access$Species=gsub("(\\w+) (\\w+)","\\1_\\2",access$Species)
rownames(access)=access$Species
mt=character() #making object for partition models
```

Only using COI, COIII and 16s.  So, I am paring the dataset down to only those genes and dropping any species for which we do not have at least one of those sequences.

```{r}
access=access[,c(1:5,7)]
access=access[!(is.na(access$COI)&is.na(access$COIII)&is.na(access$x16s)),]
sum(access$burrow=="sub-surface burrowing",na.rm=T)
sum(access$keel=="present",na.rm=T)
```

Finally, I am making a table of the accession #s used in this analysis.
```{r}
access.kable=access
colnames(access.kable)=c("Species","burrow","keel","COI","COIII","16s")
kable(access.kable,row.names = F)
```

# Multi-Gene Tree
First, I am constructing a tree using all three genes.  


## 16s
### Sequencing reading and QC
Now, I am going to repeat the process I followed with 12s with the COIII sequences with only a few modifications. Therefore, I will annotate this less. 
```{r}
x16s.accession=access$x16s
x16s.accession[is.na(x16s.accession)]="DJ078208"
x16s=read.GenBank(x16s.accession)
x16s[names(x16s)=="DJ078208"]=as.DNAbin("-")
names(x16s)=access$Species
```

Building a base frequency table.
```{r}
x16s.bases=base.freq(x16s[1])
for (i in 2:length(x16s)){
  x16s.bases=rbind(x16s.bases,base.freq(x16s[i]))  
}
rownames(x16s.bases)=names(x16s)
```


finding any possible sequences that need to be reverse complemented
```{r}
plot(x16s.bases[,1]/x16s.bases[,4],ylim=c(0,2.8))
plot(x16s.bases[,1]/x16s.bases[,4],x16s.bases[,2]/x16s.bases[,3],ylim=c(0,2.8),xlim=c(0,3))
abline(v=1,col="red")
abline(h=1,col="red")
```

These look pretty good overall, so I don't need to do much with them.


### Sequence alignment
Alignment is a lot more straight forward for these sequence since they are protein coding and not structural.  Here, I can just use the AlignSeqs function from DECIPHER package.
```{r message=FALSE, warning=FALSE, results='hide'}
x16s2=DNAbin_to_DNAstringset(x16s)
x16s.align=AlignSeqs(x16s2)
x16s.stag=StaggerAlignment(x16s.align)
x16s.final=as.DNAbin(x16s.stag)
```

### Creating nexus file
```{r}
x16s.final=as.DNAbin(x16s.stag)
write.nexus.data(x16s.final,"x16s.nxs",charsperline = 2000)
```

Finding the place to start trimming to get rid of the trailing NAs
```{r}
length(x16s.final$Abdopus_aculeatus)+max(nchar(labels(x16s.final)))+10
```

trimming off NA's

Linux version:
```{bash}
sed -i -E 's/(^.{1050}).*/\1/g' x16s.nxs
```



Then, using sed to remove the last two lines of the file
```{bash}
sed -i "$(( $(wc -l <x16s.nxs)-3+1 )),$ d" x16s.nxs
```

### Model Test
In this case, I want to partition the dataset into each of the three codon positions. That will be done in the MrBayes command block and not here. However, I do need to run model test for each codon position independently. That is what I am doing here. 
```{r}
x16s.dist=dist.dna(x16s.final[!is.na(access$x16s)])
x16s.nj=NJ(x16s.dist)
x16s.pd=as.phyDat(x16s.final[!is.na(access$x16s)])
```

```{r}
x16s.length=length(x16s$Abdopus_aculeatus)
```


x16s codon 
```{r}
start=Sys.time()
x16s.1.mT=modelTest(x16s.pd[,seq(from=1,to=x16s.length,by=3)],x16s.nj,
                        model=c("GTR","SYM","K80","HKY","F81","JC")
                        ,multicore=T,mc.cores=4
                     )
x16s.1.mT=x16s.1.mT[order(x16s.1.mT$AICc),]
x16s.1.mT$Model=gsub("\\(\\d\\)","",x16s.1.mT$Model)
x16s.1.mT$Model[1]
mt[1]=x16s.1.mT$Model[1]
save.image()
Sys.time()-start
```

## COI
### Sequencing reading and QC
Now, I am going to repeat the process I followed with 12s with the COIII sequences with only a few modifications. Therefore, I will annotate this less. 
```{r}
coi.accession=access$COI
coi.accession[is.na(coi.accession)]="DJ078208"
coi=read.GenBank(coi.accession)
coi[names(coi)=="DJ078208"]=as.DNAbin("-")
names(coi)=access$Species
```

Building a base frequency table.
```{r}
coi.bases=base.freq(coi[1])
for (i in 2:length(coi)){
  coi.bases=rbind(coi.bases,base.freq(coi[i]))  
}
rownames(coi.bases)=names(coi)
```


finding any possible sequences that need to be reverse complemented
```{r}
plot(coi.bases[,1]/coi.bases[,4],ylim=c(0,2.8))
plot(coi.bases[,1]/coi.bases[,4],coi.bases[,2]/coi.bases[,3],ylim=c(0,2.8),xlim=c(0,3))
abline(v=1,col="red")
abline(h=1,col="red")
```

These look pretty good overall, so I don't need to do much with them.


### Sequence alignment
Alignment is a lot more straight forward for these sequence since they are protein coding and not structural.  Here, I can just use the AlignSeqs function from DECIPHER package.
```{r message=FALSE, warning=FALSE, results='hide'}
coi2=DNAbin_to_DNAstringset(coi)
coi.align=AlignSeqs(coi2)
coi.stag=StaggerAlignment(coi.align)
coi.final=as.DNAbin(coi.stag)
```

### Creating nexus file
```{r}
coi.final=as.DNAbin(coi.stag)
write.nexus.data(coi.final,"coi.nxs",charsperline = 2000)
```

Finding the place to start trimming to get rid of the trailing NAs
```{r}
length(coi.final$Abdopus_aculeatus)+max(nchar(labels(coi.final)))+10
```

trimming off NA's
```{bash}
sed -i -E 's/(^.{1620}).*/\1/g' coi.nxs
```



Then, using sed to remove the last two lines of the file
```{bash}
sed -i "$(( $(wc -l <coi.nxs)-3+1 )),$ d" coi.nxs
```

### Model Test
In this case, I want to partition the dataset into each of the three codon positions. That will be done in the MrBayes command block and not here. However, I do need to run model test for each codon position independently. That is what I am doing here. 
```{r}
coi.dist=dist.dna(coi.final[!is.na(access$COI)])
coi.nj=NJ(coi.dist)
coi.pd=as.phyDat(coi.final[!is.na(access$COI)])
```

```{r}
coi.length=length(coi$Abdopus_aculeatus)
```


COI codon position 1
```{r}
start=Sys.time()
coi.1.mT=modelTest(coi.pd[,seq(from=1,to=coi.length,by=3)],coi.nj,
                        model=c("GTR","SYM","K80","HKY","F81","JC")
                        ,multicore=T,mc.cores=4
                     )
coi.1.mT=coi.1.mT[order(coi.1.mT$AICc),]
coi.1.mT$Model=gsub("\\(\\d\\)","",coi.1.mT$Model)
coi.1.mT$Model[1]
mt[2]=coi.1.mT$Model[1]
save.image()
Sys.time()-start
```

COI codon position 2
```{r}
start=Sys.time()
coi.2.mT=modelTest(coi.pd[,seq(from=2,to=coi.length,by=3)],coi.nj,
                        model=c("GTR","SYM","K80","HKY","F81","JC")
                        ,multicore=T,mc.cores=2
                     )
coi.2.mT$Model=gsub("\\(\\d\\)","",coi.2.mT$Model)
coi.2.mT=coi.2.mT[order(coi.2.mT$AICc),]
coi.2.mT$Model[1]
mt[3]=coi.2.mT$Model[1]
save.image()
Sys.time()-start
#test
```

COI codon position 3
```{r}
start=Sys.time()
coi.3.mT=modelTest(coi.pd[,seq(from=3,to=coi.length,by=3)],coi.nj,
                        model=c("GTR","SYM","K80","HKY","F81","JC")
                        ,multicore=T,mc.cores=4
                     )
coi.3.mT=coi.3.mT[order(coi.3.mT$AICc),]
coi.3.mT$Model=gsub("\\(\\d\\)","",coi.3.mT$Model)
coi.3.mT$Model[1]
mt[4]=coi.3.mT$Model[1]
save.image()
Sys.time()-start
```



## COIII
### Sequencing reading and QC
Now, I am going to repeat the process I followed with 12s with the COIII sequences with only a few modifications. Therefore, I will annotate this less. 
```{r}
coiii.accession=access$COIII
coiii.accession[is.na(coiii.accession)]="DJ078208"
coiii=read.GenBank(coiii.accession)
coiii[names(coiii)=="DJ078208"]=as.DNAbin("-")
names(coiii)=access$Species
```

Building a base frequency table.
```{r}
coiii.bases=base.freq(coiii[1])
for (i in 2:length(coiii)){
  coiii.bases=rbind(coiii.bases,base.freq(coiii[i]))  
}
rownames(coiii.bases)=names(coiii)
```


finding any possible sequences that need to be reverse complemented
```{r}
plot(coiii.bases[,1]/coiii.bases[,4],ylim=c(0,2.8))
plot(coiii.bases[,1]/coiii.bases[,4],coiii.bases[,2]/coiii.bases[,3],ylim=c(0,2.8),xlim=c(0,3))
abline(v=1,col="red")
abline(h=1,col="red")
```

These look pretty good overall, so I don't need to do much with them.


### Sequence alignment
Alignment is a lot more straight forward for these sequence since they are protein coding and not structural.  Here, I can just use the AlignSeqs function from DECIPHER package.
```{r message=FALSE, warning=FALSE, results='hide'}
coiii2=DNAbin_to_DNAstringset(coiii)
coiii.align=AlignSeqs(coiii2)
coiii.stag=StaggerAlignment(coiii.align)
coiii.final=as.DNAbin(coiii.stag)
```


### Creating nexus file
```{r}
coiii.final=as.DNAbin(coiii.stag)
write.nexus.data(coiii.final,"coiii.nxs",charsperline = 2000)
```

Finding the place to start trimming to get rid of the trailing NAs
```{r}
length(coiii.final$Abdopus_aculeatus)+max(nchar(labels(coiii.final)))+10
```
trimming off NA's

Linux version:
```{bash}
sed -i -E 's/(^.{757}).*/\1/g' coiii.nxs
```



Then, using sed to remove the last two lines of the file
```{bash}
sed -i "$(( $(wc -l <coiii.nxs)-3+1 )),$ d" coiii.nxs
```

### Model Test
In this case, I want to partition the dataset into each of the three codon positions. That will be done in the MrBayes command block and not here. However, I do need to run model test for each codon position independently. That is what I am doing here. 
```{r}
coiii.dist=dist.dna(coiii.final[!is.na(access$COIII)])
coiii.nj=NJ(coiii.dist)
coiii.pd=as.phyDat(coiii.final[!is.na(access$COIII)])
```

```{r}
coiii.length=length(coiii$Abdopus_aculeatus)
```


COIII codon position 1
```{r}
start=Sys.time()
coiii.1.mT=modelTest(coiii.pd[,seq(from=1,to=coiii.length,by=3)],coiii.nj,
                        model=c("GTR","SYM","K80","HKY","F81","JC")
                        ,multicore=T,mc.cores=4
                     )
coiii.1.mT=coiii.1.mT[order(coiii.1.mT$AICc),]
coiii.1.mT$Model=gsub("\\(\\d\\)","",coiii.1.mT$Model)
coiii.1.mT$Model[1]
mt[5]=coiii.1.mT$Model[1]
save.image()
Sys.time()-start
```

COIII codon position 2
```{r}
start=Sys.time()
coiii.2.mT=modelTest(coiii.pd[,seq(from=2,to=coiii.length,by=3)],coiii.nj,
                        model=c("GTR","SYM","K80","HKY","F81","JC")
                        ,multicore=T,mc.cores=2
                     )
coiii.2.mT$Model=gsub("\\(\\d\\)","",coiii.2.mT$Model)
coiii.2.mT=coiii.2.mT[order(coiii.2.mT$AICc),]
coiii.2.mT$Model[1]
#mt[6]=coiii.2.mT$Model[1]
mt[6]="HKY+G+I"
save.image()
Sys.time()-start
#test
```

COIII codon position 3
```{r}
start=Sys.time()
coiii.3.mT=modelTest(coiii.pd[,seq(from=3,to=coiii.length,by=3)],coiii.nj,
                        model=c("GTR","SYM","K80","HKY","F81","JC")
                        ,multicore=T,mc.cores=4
                     )
coiii.3.mT=coiii.3.mT[order(coiii.3.mT$AICc),]
coiii.3.mT$Model=gsub("\\(\\d\\)","",coiii.3.mT$Model)
coiii.3.mT$Model[1]
#mt[7]=coiii.3.mT$Model[1]
mt[7]="HKY+G"
save.image()
Sys.time()-start
```


## Models Selected
Finally, I have been saving the models selected for each parition in an object names "mt".  Here I put together that information and related information to put in a table.
```{r}
Nst=rep(6,length(mt))
Nst[grepl("K80|HKY",mt)]=2
Nst[grepl("JC|F81",mt)]=1

rate=rep("",7)
rate[grepl("\\+I",mt)]="propinv"
rate[grepl("\\+G",mt)]="gamma"
rate[grepl("\\+G\\+I",mt)]="invgamma"

models.selected=
data.frame(partition=c("16s",
                       "COXI pos 1",
                       "COXI pos 2",
                       "COXI pos 3",
                       "COXIII pos 1",
                       "COXIII pos 2",
                       "COXIII pos 3"),
           length=c(length(x16s.align),
                    floor(length(coi.align$Abdopus_aculeatus)/3+.67),
                    floor(length(coi.align$Abdopus_aculeatus)/3+.34),
                    floor(length(coi.align$Abdopus_aculeatus)/3),
                    floor(length(coiii.align$Abdopus_aculeatus)/3+.67),
                    floor(length(coiii.align$Abdopus_aculeatus)/3+.34),
                    floor(length(coiii.align$Abdopus_aculeatus)/3)),

           model=mt,
           nucmodel=c(rep("4by4",7)),
           Nst,
           rate)
models.selected$model=gsub("\\+G","+Γ",models.selected$model)
write.csv(models.selected,"models_selected.csv",row.names = F)
kable(models.selected,align=c("l","c","c","c","c","r"))
```

## Combine nexus file production
THe final step of the data preparation it so make a combine nexus file of all three genes and the MrBayes command block. 
First, I am writing the 12s stem and loops positions out to a text file we can incorportate into the nexus file later.
```{r}#
write.table(t(stems.12s),"stems12s",sep=" ",row.names = F,col.names = F)
write.table(t(loops.12s),"loops12s",sep=" ",row.names = F,col.names = F)
```

Next, I find the length of the whole dataset
```{r}
total.len=length(x16s.final$Abdopus_aculeatus)+
  length(coi.final$Abdopus_aculeatus)+
  length(coiii.final$Abdopus_aculeatus)
total.len
```
The start of the coi dataset:
```{r}
coi.start=length(x16s.final$Abdopus_aculeatus)+1
coi.start
```
The start of the coiii dataset:
```{r}
coiii.start=length(x16s.final$Abdopus_aculeatus)+
  length(coiii.final$Abdopus_aculeatus)+1
coiii.start
```

### Making the model selection block
Next, I construct the part of the MrBayes command block in which I will apply the settings for each model selected.  I am using a look-up table of settings I produced and stored in the "model_selection.csv" file. I then write this part of the command block out to a text file that I will incorportate into the combine nexus file later.
```{r}
models=read.csv("model_selection.csv")
models.muus=character()
for (i in 1:length(mt)){
  models.muus[i]=paste("      lset applyto=(",i,") ",models$lset[models$model==mt[i]],";",sep="")
}
for (i in 1:length(mt)){
  models.muus[i+length(mt)]=paste("      prset applyto=(",i,") ",models$prset[models$model==mt[i]],";",sep="")
}
models.muus=models.muus[-grep("NA",models.muus)]
write.table(models.muus,"models_to_write",row.names = F,col.names = F,quote = F)
```


### Making blocks that require lengths
Here I am making various blocks that require sequence lengths.  That includes the header of the nexux file and the portion of the MrBayes command block paritioning the dataset.
```{r}
write.table(paste("  DIMENSIONS NTAX=",nrow(access)," NCHAR=",total.len,";",sep=""),"dimensions",
            row.names = F,col.names = F,quote = F)

write.table(data.frame(models=c(
            paste("      charset 16s =  ",1," - ",coi.start-1,";",sep=""))),
            "charset.16s",
            row.names = F,col.names = F,quote = F)

write.table(data.frame(models=c(
            paste("      charset coi_pos1 =  ",coi.start," - ",coiii.start-1,"\\3;",sep=""),
            paste("      charset coi_pos2 =  ",coi.start+1," - ",coiii.start-1,"\\3;",sep=""),
            paste("      charset coi_pos3 =  ",coi.start+2," - ",coiii.start-1,"\\3;",sep="")
            )),
            "charset.coi",
            row.names = F,col.names = F,quote = F)

write.table(data.frame(models=c(
            paste("      charset coiii_pos1 =  ",coiii.start," - ",total.len,"\\3;",sep=""),
            paste("      charset coiii_pos2 =  ",coiii.start+1," - ",total.len,"\\3;",sep=""),
            paste("      charset coiii_pos3 =  ",coiii.start+2," - ",total.len,"\\3;",sep="")
            )),
            "charset.coiii",
            row.names = F,col.names = F,quote = F)

```


### Generating nexus file with MrBayes command block
Finally, we can take all of those text files we produced and use some BASH to create the nexus file that we will use in the analysis. It will be called "complete.nxs".  
```{bash, eval=F}
rm complete.nxs #just cleaning out previous iteration if running multiple times
cp -r mb mb.backup
rm -r mb
touch complete.nxs
echo "#NEXUS" >> complete.nxs
echo "BEGIN DATA;" >> complete.nxs
sed -n 1p dimensions >> complete.nxs
echo "  FORMAT DATATYPE=DNA MISSING=? GAP=- INTERLEAVE=YES;
  MATRIX
  
  [16s]" >> complete.nxs 
sed -n 7,116p x16s.nxs >> complete.nxs
echo "
   [COI]" >> complete.nxs
sed -n 7,116p coi.nxs >> complete.nxs
echo "
   [COIII]" >> complete.nxs
sed -n 7,116p coiii.nxs >> complete.nxs
echo "  ;
END;

begin mrbayes;" >> complete.nxs
echo " " >> complete.nxs

sed -n 1,3p charset.16s >> complete.nxs
sed -n 1,3p charset.coi >> complete.nxs
sed -n 1,3p charset.coiii >> complete.nxs
echo " " >> complete.nxs
echo "      partition by_gene = 7:16s,coi_pos1,coi_pos2,coi_pos3,coiii_pos1,coiii_pos2,coiii_pos3;" >> complete.nxs
echo "	    set partition = by_gene;
" >> complete.nxs
#echo "	    constraint outg = Octopus_vulgaris Octopus_rubescens Octopus_bimaculoides Octopus_cyanea Amphioctopus_aegina Hapalochlaena_maculosa Abdopus_aculeatus Octopus_tetricus Octopus_berrima Macroctopus_maorum;
#	    prset topologypr = constraints (outg);
#	    outgroup Octopus_vulgaris;
#	    " >> complete.nxs
echo "	    lset applyto=(1,2,3,4,5,6,7)   nucmodel=4by4;
	    prset ratepr=variable;
	    unlink statefreq=(all) revmat=(all) shape=(all) pinvar=(all);" >> complete.nxs

cat models_to_write >> complete.nxs

echo "
      mcmcp ngen=100000000 printfreq=1000 samplefreq=10000 stoprule=yes stopval=0.01 nruns=5 nchains=2 burninfrac=0.25;
      mcmc;
      sump;
      sumt filename=mb/complete.nxs contype=allcompat conformat=simple;

end;
" >> complete.nxs

mkdir mb
cp complete.nxs mb/complete.nxs
```

## Running MrBayes analysis
Now, we run the analysis with this BASH command.  I have compiled MrBayes from source with MPI multi-threading enabled.
```{bash, eval=F}
mpiexec -np 10 mb mb/complete.nxs > mb_log
```

# Reading into R the resulting trees

## Reading in the multi-gene tree
The file that contains the consensus trees is "complete.nxs.con.tre". This file actually contains two trees.  One contains no branch probability information, and the other contains the branch probabilities.  The following code deletes the tree without probabilities from file. If both trees are present, R opens the tree as a multi-phylo class, and manipulation of the tree becomes much harder.
```{bash eval=FALSE}
cp ./mb/complete.nxs.con.tre ./mb/completeBACKUP.nxs.con.tre
sed -i '/Note: This tree contains information only on the topology/d' ./mb/complete.nxs.con.tre
sed -i '/and branch lengths (median of the posterior probability density)/d' ./mb/complete.nxs.con.tre
sed -i '/):/d' ./mb/complete.nxs.con.tre
```


Now I read the tree into R using the following code.
```{r}
mrbayes.tree=read.nexus("mb/complete.nxs.con.tre",tree.names=NULL)
```

```{r}
mrbayes.tree$node.label
```

```{r}
root.node=
  getMRCA(mrbayes.tree,tip=c("Stauroteuthis_gilchristi","Cirrothauma_murrayi","Opisthoteuthis_depressa","Cirroctopus_glacialis"))
```

```{r}
mrbayes.root=root(mrbayes.tree,node=root.node)
```


# Pagel's Test

```{r}
unique(access$burrow)
access$burrow[access$burrow=="sub-surface burrowing"]="burying"
access$burrow[access$burrow=="unknown"]="non-burrowing"
```

```{r}
unique(access$keel)
access$keel[access$keel=="unknown"]="absent"
```

```{r}
burrow=setNames(access$burrow,access$Species)
keel=setNames(access$keel,access$Species)



```

```{r}
fit.keeldep=fitPagel(mrbayes.root,burrow,keel,dep.var="y")
fit.burydep=fitPagel(mrbayes.root,burrow,keel,dep.var="x")
fit.interdep=fitPagel(mrbayes.root,burrow,keel,dep.var="xy")
```

```{r}
as.numeric(fit.keeldep$dependent.AIC)
as.numeric(fit.burydep$dependent.AIC)
as.numeric(fit.interdep$dependent.AIC)
as.numeric(fit.interdep$independent.AIC)

```
```{r}
fit.interdep
```


```{r}
svg("pagel_test.svg")

plot(fit.interdep,cex=0.5,lwd.by.rate=T)
dev.off()
```


# Plotting the tree

```{r}
svg(file="Figure2.svg",width=8,height=12)
plot(mrbayes.root,show.node.label=F,use.edge.length=T,align.tip.label=F,edge.width=1,
     show.tip.label = T,x.lim=1.65,font=3,cex=0.5)
#nodelabels(1:108, adj = c(0.5, 0.5), frame = "n",
#            cex = 0.5,font=2)
dev.off()
```



This is bash code to convert the svg to png that can be displayed in the RMarkdown pdf.
```{bash}
cairosvg Figure2.svg -o Figure2.png -d 300
```

And this is bash code to convert the svg to eps for publication.
```{bash}
inkscape Figure2.svg -o Figure2.eps --export-ignore-filters --export-ps-level=3
```


![Phylogeny of family Enteroctopodidae](Figure2.png)


```{r}
mrbayes.root$tip.label %in% access$Species 

cbind(access$Species,mrbayes.root$tip.label)
```



```{r}
keel.col=rep("grey",110)
keel.col[access$keel=="present"]="#0cceffff"
bury.col=rep("grey",110)
bury.col[access$burrow=="burying"]="#0cceffff"
```


# Plotting the
```{r}
par(mfrow=c(1,2))
plot(mrbayes.root,show.tip.label=FALSE,no.margin=TRUE,use.edge.length=F)
tiplabels("●",frame="n",cex=0.3,adj=c(0,0.3),col=keel.col)
plot(mrbayes.root,show.tip.label=FALSE,no.margin=TRUE,direction="leftwards",use.edge.length=F)
tiplabels("●",frame="n",cex=0.3,adj=c(0.7,0.3),col=bury.col)


```

```{r}
f = function(n) c("grey", "red")
g = function(n) c("red", "grey")
```

```{r}
svg("character_phylo.svg",height=10,width=10)
par(mar=c(0, 4, 4, 2))
plot(mrbayes.root,show.tip.label=FALSE,use.edge.length=F)
#tiplabels("●",frame="n",cex=0.5,adj=c(-1,0.25),col=keel.col)
#tiplabels("●",frame="n",cex=0.5,adj=c(-3,0.25),col=bury.col)
phydataplot(setNames(access$burrow,access$Species),mrbayes.root,style="mosaic",width=3,legend="none",funcol=g)
phydataplot(setNames(access$keel,access$Species),mrbayes.root,style="mosaic",width=3,offset=5,legend="none",funcol = f)
#ring(setNames(rep(1:110),access$Species),mrbayes.root,style="a")
dev.off()
```
```{r}
rainbow(2)
```


```{r}
keel.col=rep("#ccf4ffff",110)
keel.col[access$keel=="present"]="#0cceffff"
bury.col=rep("#ffd6ccff",110)
bury.col[access$burrow=="burying"]="#FF3D0C"

svg("ring_phylo.svg",height=10,width=11)
par(mar=c(2, 1, 2, 7)+5)
par(xpd=T)
plot(mrbayes.root, "f", show.tip.label = FALSE, x.l = c(-1, 1), y.l = c(-1, 1),use.edge.length=F,edge.width=0)
ring(0.2, mrbayes.root, "r", col = keel.col, offset = 0)
ring(0.2, mrbayes.root, "r", col = bury.col, offset = 0.22)
par(new=T)
plot(mrbayes.root, "f", x.l = c(-1, 1), y.l = c(-1, 1),use.edge.length=F,edge.width=2,cex=0.7)
par(new=T)
par(mar=c(0, 0, 0, 0))
plot(c(0,1),c(0,1),axes=F, ylab="",xlab="",type="n")
legend("bottomright",c("has keel","does not have keel","known to burrow or bury","not known to burrow or bury"),fill=c("#0cceffff","#ccf4ffff","#FF3D0C","#ffd6ccff"),border=c("#0cceffff","#ccf4ffff","#FF3D0C","#ffd6ccff"),bty="n",cex=1.3)
dev.off()

```
This is bash code to convert the svg to png that can be displayed in the RMarkdown pdf.
```{bash}
cairosvg ring_phylo.svg -o ring_phylo.png -d 300
```

