---
title: "Burrows Bay Octopus Phylogeny, edited for Cheyne's Thesis"
author: "Kirt Onthank"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_depth: '3'
    df_print: paged
  pdf_document:
    toc: yes
    number_sections: yes
    toc_depth: 3
---

# Loading Libraries
```{r Libraries, message=FALSE, warning=FALSE}
library(ape)
library(xlsx)
library(insect)
library(aphid)
library(DECIPHER)
library(magrittr)
library(phangorn)
library(rwty)
library(stringi)
library(Biostrings)
library(kableExtra)
library(knitr)
```


This function to convert DNAbin objects used by the ape package to DNAStringSet objects used by the DECIPHER package was written by Joel Nitta, available on his github (https://gist.github.com/joelnitta/6f30a7c0f1c83d78c76a5469e935d56f)
```{r}
# Convert ape::DNAbin format to Biostrings::DNAStringSet format,
# optionally removing gaps
DNAbin_to_DNAstringset <- function (seqs, remove_gaps = TRUE) {
  if(isTRUE(remove_gaps)) {
  seqs %>% as.list() %>% as.character %>% 
      lapply(.,paste0,collapse="") %>% 
      lapply( function (x) gsub("-", "", x)) %>% 
      unlist %>% Biostrings::DNAStringSet()
  } else {
    seqs %>% as.list() %>% as.character %>% 
      lapply(.,paste0,collapse="") %>% 
      unlist %>% Biostrings::DNAStringSet()
  }
}
```


# Preparing the dataset
## Reading in accession numbers
First I read in the accession numbers and 
```{r}
access=read.csv("Octopus burrowing - data.csv",na.strings = "")
colnames(access)=c("Species","burrow","keel","reference","COI","COIII","x12s","x16s","x28s","Rhodopsin","Pax-6","cytb","notes")
access=access[,1:13] #Dropping the junk columns
access=access[,-c(4,13)] #Dropping notes and references columns
mt=character() #making object for partition models
```


I am dropping all of the taxa for which sequences for all three genes (COIII, CytB and 12s) are available.  The one exception is Octopus rubescens, which some reviewers have asked me to include so that I could exclude both octopus species that occur in the same area.
```{r}
#access=access[complete.cases(access)|access$NA.=="Octopus_rubescens",] 
```

Next, I am limiting the dataset to only a single set of sequences for each species or subspecies.
```{r}
#access=access[!grepl("_\\d",access$NA.),]
```

Finally, I am making a table of the accession #s used in this analysis.
```{r}
access.kable=access
colnames(access.kable)=c("Species","burrow","keel","COI","COIII","x12s","x16s","x28s","Rhodopsin","Pax-6","cytb")
kable(access.kable,row.names = F)
```

# Multi-Gene Tree
First, I am constructing a tree using all three genes.  


## 16s
### Sequencing reading and QC
Now, I am going to repeat the process I followed with 12s with the COIII sequences with only a few modifications. Therefore, I will annotate this less. 
```{r}
x16s.accession=access$x16s
x16s.accession[is.na(x16s.accession)]="DJ078208"
x16s=read.GenBank(x16s.accession)
x16s[names(x16s)=="DJ078208"]=as.DNAbin("-")
names(x16s)=access$Species
```

Building a base frequency table.
```{r}
x16s.bases=base.freq(x16s[1])
for (i in 2:length(x16s)){
  x16s.bases=rbind(x16s.bases,base.freq(x16s[i]))  
}
rownames(x16s.bases)=names(x16s)
```


finding any possible sequences that need to be reverse complemented
```{r}
plot(x16s.bases[,1]/x16s.bases[,4],ylim=c(0,2.8))
plot(x16s.bases[,1]/x16s.bases[,4],x16s.bases[,2]/coi.bases[,3],ylim=c(0,2.8),xlim=c(0,3))
abline(v=1,col="red")
abline(h=1,col="red")
```

These look pretty good overall, so I don't need to do much with them.


### Sequence alignment
Alignment is a lot more straight forward for these sequence since they are protein coding and not structural.  Here, I can just use the AlignSeqs function from DECIPHER package.
```{r message=FALSE, warning=FALSE, results='hide'}
x16s2=DNAbin_to_DNAstringset(x16s)
x16s.align=AlignSeqs(x16s2)
x16s.stag=StaggerAlignment(x16s.align)
x16s.final=as.DNAbin(x16s.stag)
```

### Creating nexus file
```{r}
x16s.final=as.DNAbin(x16s.stag)
write.nexus.data(x16s.final,"x16s.nxs",charsperline = 1000)
```

Finding the place to start trimming to get rid of the trailing NAs
```{r}
length(x16s.final$Abdopus_aculeatus)+max(nchar(labels(x16s.final)))+10
```

trimming off NA's

Linux version:
```{bash}
sed -i -E 's/(^.{1004}).*/\1/g' x16s.nxs
```



Then, using sed to remove the last two lines of the file
```{bash}
sed -i "$(( $(wc -l <x16s.nxs)-3+1 )),$ d" x16s.nxs
```

### Model Test
In this case, I want to partition the dataset into each of the three codon positions. That will be done in the MrBayes command block and not here. However, I do need to run model test for each codon position independently. That is what I am doing here. 
```{r}
x16s.dist=dist.dna(x16s.final[!is.na(access$x16s)])
x16s.nj=NJ(x16s.dist)
x16s.pd=as.phyDat(x16s.final[!is.na(access$x16s)])
```

```{r}
x16s.length=length(x16s$Abdopus_aculeatus)
```


x16s codon 
```{r}
start=Sys.time()
x16s.1.mT=modelTest(x16s.pd[,seq(from=1,to=x16s.length,by=3)],x16s.nj,
                        model=c("GTR","SYM","K80","HKY","F81","JC")
#                        ,multicore=T,mc.cores=4
                     )
x16s.1.mT=x16s.1.mT[order(x16s.1.mT$AICc),]
x16s.1.mT$Model=gsub("\\(\\d\\)","",x16s.1.mT$Model)
x16s.1.mT$Model[1]
mt[3]=x16s.1.mT$Model[1]
save.image()
Sys.time()-start
```

## COI
### Sequencing reading and QC
Now, I am going to repeat the process I followed with 12s with the COIII sequences with only a few modifications. Therefore, I will annotate this less. 
```{r}
coi.accession=access$COI
coi.accession[is.na(coi.accession)]="DJ078208"
coi=read.GenBank(coi.accession)
coi[names(coi)=="DJ078208"]=as.DNAbin("-")
names(coi)=access$Species
```

Building a base frequency table.
```{r}
coi.bases=base.freq(coi[1])
for (i in 2:length(coi)){
  coi.bases=rbind(coi.bases,base.freq(coi[i]))  
}
rownames(coi.bases)=names(coi)
```


finding any possible sequences that need to be reverse complemented
```{r}
plot(coi.bases[,1]/coi.bases[,4],ylim=c(0,2.8))
plot(coi.bases[,1]/coi.bases[,4],coi.bases[,2]/coi.bases[,3],ylim=c(0,2.8),xlim=c(0,3))
abline(v=1,col="red")
abline(h=1,col="red")
```

These look pretty good overall, so I don't need to do much with them.


### Sequence alignment
Alignment is a lot more straight forward for these sequence since they are protein coding and not structural.  Here, I can just use the AlignSeqs function from DECIPHER package.
```{r message=FALSE, warning=FALSE, results='hide'}
coi2=DNAbin_to_DNAstringset(coi)
coi.align=AlignSeqs(coi2)
coi.stag=StaggerAlignment(coi.align)
coi.final=as.DNAbin(coi.stag)
```

### Creating nexus file
```{r}
coi.final=as.DNAbin(coi.stag)
write.nexus.data(coi.final,"coi.nxs",charsperline = 1000)
```

Finding the place to start trimming to get rid of the trailing NAs
```{r}
length(coi.final$Abdopus_aculeatus)+max(nchar(labels(coi.final)))+10
```

trimming off NA's

Linux version:
```{bash}
sed -i -E 's/(^.{1620}).*/\1/g' coi.nxs
```



Then, using sed to remove the last two lines of the file
```{bash}
sed -i "$(( $(wc -l <coi.nxs)-3+1 )),$ d" coi.nxs
```

### Model Test
In this case, I want to partition the dataset into each of the three codon positions. That will be done in the MrBayes command block and not here. However, I do need to run model test for each codon position independently. That is what I am doing here. 
```{r}
coi.dist=dist.dna(coi.final[!is.na(access$COI)])
coi.nj=NJ(coi.dist)
coi.pd=as.phyDat(coi.final[!is.na(access$COI)])
```

```{r}
coi.length=length(coi$Abdopus_aculeatus)
```


COI codon position 1
```{r}
start=Sys.time()
coi.1.mT=modelTest(coi.pd[,seq(from=1,to=coi.length,by=3)],coi.nj,
                        model=c("GTR","SYM","K80","HKY","F81","JC")
#                        ,multicore=T,mc.cores=4
                     )
coi.1.mT=coi.1.mT[order(coi.1.mT$AICc),]
coi.1.mT$Model=gsub("\\(\\d\\)","",coi.1.mT$Model)
coi.1.mT$Model[1]
mt[3]=coi.1.mT$Model[1]
save.image()
Sys.time()-start
```

COI codon position 2
```{r}
start=Sys.time()
coi.2.mT=modelTest(coi.pd[,seq(from=2,to=coi.length,by=3)],coi.nj,
                        model=c("GTR","SYM","K80","HKY","F81","JC")
#                        ,multicore=T,mc.cores=2
                     )
coi.2.mT$Model=gsub("\\(\\d\\)","",coi.2.mT$Model)
coi.2.mT=coi.2.mT[order(coi.2.mT$AICc),]
coi.2.mT$Model[1]
mt[4]=coi.2.mT$Model[1]
save.image()
Sys.time()-start
#test
```

COI codon position 3
```{r}
start=Sys.time()
coi.3.mT=modelTest(coi.pd[,seq(from=3,to=coi.length,by=3)],coi.nj,
                        model=c("GTR","SYM","K80","HKY","F81","JC")
   #                     ,multicore=T,mc.cores=4
                     )
coi.3.mT=coi.3.mT[order(coi.3.mT$AICc),]
coi.3.mT$Model=gsub("\\(\\d\\)","",coi.3.mT$Model)
coi.3.mT$Model[1]
mt[5]=coi.3.mT$Model[1]
save.image()
Sys.time()-start
```



## COIII
### Sequencing reading and QC
Now, I am going to repeat the process I followed with 12s with the COIII sequences with only a few modifications. Therefore, I will annotate this less. 
```{r}
coiii.accession=access$COIII
coiii.accession[is.na(coiii.accession)]="DJ078208"
coiii=read.GenBank(coiii.accession)
coiii[names(coiii)=="DJ078208"]=as.DNAbin("-")
names(coiii)=access$Species
```

Building a base frequency table.
```{r}
coiii.bases=base.freq(coiii[1])
for (i in 2:length(coiii)){
  coiii.bases=rbind(coiii.bases,base.freq(coiii[i]))  
}
rownames(coiii.bases)=names(coiii)
```


finding any possible sequences that need to be reverse complemented
```{r}
plot(coiii.bases[,1]/coiii.bases[,4],ylim=c(0,2.8))
plot(coiii.bases[,1]/coiii.bases[,4],coiii.bases[,2]/coiii.bases[,3],ylim=c(0,2.8),xlim=c(0,3))
abline(v=1,col="red")
abline(h=1,col="red")
```

These look pretty good overall, so I don't need to do much with them.


### Sequence alignment
Alignment is a lot more straight forward for these sequence since they are protein coding and not structural.  Here, I can just use the AlignSeqs function from DECIPHER package.
```{r message=FALSE, warning=FALSE, results='hide'}
coiii2=DNAbin_to_DNAstringset(coiii)
coiii.align=AlignSeqs(coiii2)
coiii.stag=StaggerAlignment(coiii.align)
coiii.final=as.DNAbin(coiii.stag)
```


### Creating nexus file
```{r}
coiii.final=as.DNAbin(coiii.stag)
write.nexus.data(coiii.final,"coiii.nxs",charsperline = 1000)
```

Finding the place to start trimming to get rid of the trailing NAs
```{r}
length(coiii.final$Abdopus_aculeatus)+max(nchar(labels(coiii.final)))+10
```
trimming off NA's

Linux version:
```{bash}
sed -i -E 's/(^.{757}).*/\1/g' coiii.nxs
```



Then, using sed to remove the last two lines of the file
```{bash}
sed -i "$(( $(wc -l <coiii.nxs)-3+1 )),$ d" coiii.nxs
```

### Model Test
In this case, I want to partition the dataset into each of the three codon positions. That will be done in the MrBayes command block and not here. However, I do need to run model test for each codon position independently. That is what I am doing here. 
```{r}
coiii.dist=dist.dna(coiii.final[!is.na(access$COIII)])
coiii.nj=NJ(coiii.dist)
coiii.pd=as.phyDat(coiii.final[!is.na(access$COIII)])
```

```{r}
coiii.length=length(coiii$Abdopus_aculeatus)
```


COIII codon position 1
```{r}
start=Sys.time()
coiii.1.mT=modelTest(coiii.pd[,seq(from=1,to=coiii.length,by=3)],coiii.nj,
                        model=c("GTR","SYM","K80","HKY","F81","JC")
#                        ,multicore=T,mc.cores=4
                     )
coiii.1.mT=coiii.1.mT[order(coiii.1.mT$AICc),]
coiii.1.mT$Model=gsub("\\(\\d\\)","",coiii.1.mT$Model)
coiii.1.mT$Model[1]
mt[3]=coiii.1.mT$Model[1]
save.image()
Sys.time()-start
```

COIII codon position 2
```{r}
start=Sys.time()
coiii.2.mT=modelTest(coiii.pd[,seq(from=2,to=coiii.length,by=3)],coiii.nj,
                        model=c("GTR","SYM","K80","HKY","F81","JC")
#                        ,multicore=T,mc.cores=2
                     )
coiii.2.mT$Model=gsub("\\(\\d\\)","",coiii.2.mT$Model)
coiii.2.mT=coiii.2.mT[order(coiii.2.mT$AICc),]
coiii.2.mT$Model[1]
mt[4]=coiii.2.mT$Model[1]
save.image()
Sys.time()-start
#test
```

COIII codon position 3
```{r}
start=Sys.time()
coiii.3.mT=modelTest(coiii.pd[,seq(from=3,to=coiii.length,by=3)],coiii.nj,
                        model=c("GTR","SYM","K80","HKY","F81","JC")
   #                     ,multicore=T,mc.cores=4
                     )
coiii.3.mT=coiii.3.mT[order(coiii.3.mT$AICc),]
coiii.3.mT$Model=gsub("\\(\\d\\)","",coiii.3.mT$Model)
coiii.3.mT$Model[1]
mt[5]=coiii.3.mT$Model[1]
save.image()
Sys.time()-start
```


## Models Selected
Finally, I have been saving the models selected for each parition in an object names "mt".  Here I put together that information and related information to put in a table.
```{r}
Nst=rep(6,length(mt))
Nst[grepl("K80|HKY",mt)]=2
Nst[grepl("JC|F81",mt)]=1

rate=rep("",8)
rate[grepl("\\+I",mt)]="propinv"
rate[grepl("\\+G",mt)]="gamma"
rate[grepl("\\+G\\+I",mt)]="invgamma"

models.selected=
data.frame(partition=c("16s stems",
                       "16s loops",
                       "COXIII pos 1",
                       "COXIII pos 2",
                       "COXIII pos 3",
                       "COXI pos 1",
                       "COXI pos 2",
                       "COXI pos 3"),
           length=c(length(stems.12s),
                    length(loops.12s),
                    floor(length(coiii.align$Octopus_vulgaris)/3+.67),
                    floor(length(coiii.align$Octopus_vulgaris)/3+.34),
                    floor(length(coiii.align$Octopus_vulgaris)/3),
                    floor(length(cytb.align$Octopus_vulgaris)/3+.67),
                    floor(length(cytb.align$Octopus_vulgaris)/3+.34),
                    floor(length(cytb.align$Octopus_vulgaris)/3)),
           model=mt,
           nucmodel=c("doublet",rep("4by4",7)),
           Nst,
           rate)
models.selected$model=gsub("\\+G","+Γ",models.selected$model)
write.csv(models.selected,"models_selected.csv",row.names = F)
kable(models.selected,align=c("l","c","c","c","c","r"))
```

## Combine nexus file production
THe final step of the data preparation it so make a combine nexus file of all three genes and the MrBayes command block. 
First, I am writing the 12s stem and loops positions out to a text file we can incorportate into the nexus file later.
```{r}
write.table(t(stems.12s),"stems12s",sep=" ",row.names = F,col.names = F)
write.table(t(loops.12s),"loops12s",sep=" ",row.names = F,col.names = F)
```

Next, I find the length of the whole dataset
```{r}
total.len=length(x12s.final[1,])+
  length(coiii.final$Octopus_vulgaris)+
  length(cytb.final$Octopus_vulgaris)
total.len
```
The start of the coiii dataset:
```{r}
coiii.start=length(x12s.final[1,])+1
coiii.start
```
The start of the cytb dataset:
```{r}
cytb.start=length(x12s.final[1,])+
  length(coiii.final$Octopus_vulgaris)+1
cytb.start
```

### Making the model selection block
Next, I construct the part of the MrBayes command block in which I will apply the settings for each model selected.  I am using a look-up table of settings I produced and stored in the "model_selection.csv" file. I then write this part of the command block out to a text file that I will incorportate into the combine nexus file later.
```{r}
models=read.csv("model_selection.csv")
models.muus=character()
for (i in 1:length(mt)){
  models.muus[i]=paste("      lset applyto=(",i,") ",models$lset[models$model==mt[i]],";",sep="")
}
for (i in 1:length(mt)){
  models.muus[i+length(mt)]=paste("      prset applyto=(",i,") ",models$prset[models$model==mt[i]],";",sep="")
}
models.muus=models.muus[-grep("NA",models.muus)]
write.table(models.muus,"models_to_write",row.names = F,col.names = F,quote = F)
```

### Making the 12s stem base pair block
Now, I am also making a text file that has the 12s stem base pairs.
```{r}
write.table(t(c(as.vector(t(read.csv("basepairs_12s.csv",header=F)[1,])))),"basepairs",row.names = F,col.names = F)
```

And now doing a bit of reformatting of the text file using sed.
```{bash}
sed -i -E 's/([0-9]{2,4}) ([0-9]{2,4})/\1:\2/g' basepairs
```



### Making blocks that require lengths
Here I am making various blocks that require sequence lengths.  That includes the header of the nexux file and the portion of the MrBayes command block paritioning the dataset.
```{r}
write.table(paste("  DIMENSIONS NTAX=",nrow(access)," NCHAR=",total.len,";",sep=""),"dimensions",
            row.names = F,col.names = F,quote = F)

write.table(data.frame(models=c(
            paste("      charset coiii_pos1 =  ",coiii.start," - ",cytb.start-1,"\\3;",sep=""),
            paste("      charset coiii_pos2 =  ",coiii.start+1," - ",cytb.start-1,"\\3;",sep=""),
            paste("      charset coiii_pos3 =  ",coiii.start+2," - ",cytb.start-1,"\\3;",sep="")
            )),
            "charset.coii",
            row.names = F,col.names = F,quote = F)

write.table(data.frame(models=c(
            paste("      charset cytb_pos1 =  ",cytb.start," - ",total.len,"\\3;",sep=""),
            paste("      charset cytb_pos2 =  ",cytb.start+1," - ",total.len,"\\3;",sep=""),
            paste("      charset cytb_pos3 =  ",cytb.start+2," - ",total.len,"\\3;",sep="")
            )),
            "charset.cytb",
            row.names = F,col.names = F,quote = F)

```


### Generating nexus file with MrBayes command block
Finally, we can take all of those text files we produced and use some BASH to create the nexus file that we will use in the analysis. It will be called "complete.nxs".  
```{bash, eval=F}
rm complete.nxs #just cleaning out previous iteration if running multiple times
cp -r mb mb.backup
rm -r mb
touch complete.nxs
echo "#NEXUS" >> complete.nxs
echo "BEGIN DATA;" >> complete.nxs
sed -n 1p dimensions >> complete.nxs
echo "  FORMAT DATATYPE=DNA MISSING=? GAP=- INTERLEAVE=YES;
  MATRIX
  
  [12s]" >> complete.nxs 
sed -n 7,96p 12s.nxs >> complete.nxs
echo "
   [COIII]" >> complete.nxs
sed -n 7,96p coiii.nxs >> complete.nxs
echo "
   [Cytb]" >> complete.nxs
sed -n 7,98p cytb.nxs >> complete.nxs
echo "  ;
END;

begin mrbayes;
	[Define pairs for the doublet model]" >> complete.nxs
cat basepairs | sed '1s/^/    pairs /' | sed -E 's/(:[0-9]{,4})/\1,/g' | sed -E 's/,$/;/g' | sed -r 's/(.{72,76} )/\1\n         /g' >> complete.nxs
echo " " >> complete.nxs
cat stems12s | sed '1s/^/    	charset  12s-stems = /' | sed -E 's/$/;/g' | sed -r 's/(.{73,76} )/\1\n           /g' | sed '1s/^/\n/' >> complete.nxs

echo " " >> complete.nxs
cat loops12s | sed '1s/^/    	charset  12s-loops = /' | sed -E 's/$/;/g' | sed -r 's/(.{73,76} )/\1\n           /g' | sed '1s/^/\n/' >> complete.nxs
echo " " >> complete.nxs
sed -n 1,3p charset.coii >> complete.nxs
sed -n 1,3p charset.cytb >> complete.nxs
echo " " >> complete.nxs
echo "      partition by_gene = 8:12s-stems,12s-loops,coiii_pos1,coiii_pos2,coiii_pos3,cytb_pos1,cytb_pos2,cytb_pos3;" >> complete.nxs
echo "	    set partition = by_gene;
" >> complete.nxs
echo "	    constraint outg = Octopus_vulgaris Octopus_rubescens Octopus_bimaculoides Octopus_cyanea Amphioctopus_aegina Hapalochlaena_maculosa Abdopus_aculeatus Octopus_tetricus Octopus_berrima Macroctopus_maorum;
	    prset topologypr = constraints (outg);
	    outgroup Octopus_vulgaris;
	    " >> complete.nxs
echo "	    lset applyto=(1)   nucmodel=doublet;
	    lset applyto=(2,3,4,6,7,8)   nucmodel=4by4;
	    prset ratepr=variable;
	    unlink statefreq=(all) revmat=(all) shape=(all) pinvar=(all);" >> complete.nxs

cat models_to_write >> complete.nxs

echo "
      mcmcp ngen=100000000 printfreq=1000 samplefreq=10000 stoprule=yes stopval=0.01 nruns=5 nchains=2 burninfrac=0.25;
      mcmc;
      sump;
      sumt filename=mb/complete.nxs contype=allcompat conformat=simple;

end;
" >> complete.nxs

mkdir mb
cp complete.nxs mb/complete.nxs
```

## Running MrBayes analysis
Now, we run the analysis with this BASH command.  I have compiled MrBayes from source with MPI multi-threading enabled.
```{bash, eval=F}
mpiexec -np 10 mb mb/complete.nxs > mb_log
```


## Assessing convergency with RWTY
The R package RWTY (Standing for R We There Yet), is useful for determining if a Bayesian phylogenetic analysis has run for long enough.  Specifically it produces visualizations that help assess if the MCMC chains have converged. 

```{r, eval=F}
octo.trees=load.multi("mb/", format = "mb")
check.chains(octo.trees)
```

```{r, eval=F}
rwty.processors <<- 15
octo.rwty=analyze.rwty(octo.trees,filename="Burrows_Bay_octo_rwty.pdf")
```


### Log likelihood plots
First, we will take a look at the log likelihood traces for the 5 different runs.  We expect these to have no trend in an analysis that has converged. In addition, estimated sample size (ESS) should be above 200 for each parameter. As with most of the graphs in this analysis, trace plots show the change in some factor (y-axis) over successive generations of the analysis (x-axis). This allows you to see if the factor was changing, and in what way, over the course of the analysis. The density plot will show the distribution of that factor over all generations, allowing easy comparison between runs.

```{r RWTY parameter plot, eval=F}
png("LnL_trace.png",width=2000,height=1000,res=200)
makeplot.param(octo.trees, burnin = 50, "LnL")$trace.plot
dev.off()

png("LnL_density.png",width=2000,height=1000,res=200)
makeplot.param(octo.trees, burnin = 50, "LnL")$density.plot
dev.off()

```
![Log likelihood trace](LnL_trace.png)

In the density plot we are looking to see that the values sampled by each of the runs is approximately the same.

![Log likelihood density plot](LnL_density.png)

### Tree topology plot
Tree topology trace show the difference in sampled tree topology from the focal tree (on the y-axis) across generations of the bayesian analysis (x-axis). According the documentation: "Well mixed chains will show the topology trace jumping rapidly between values. Poorly mixed chains will show chains remaining at similar values in successive samples. Topology traces can also reveal whether different chains are sampling different bits of tree space, in which case the stationary distributions will show traces at different y values on the plot." (https://cran.r-project.org/web/packages/rwty/vignettes/rwty.html)

```{r RWTY topology plot, eval=F}
png("Topology_trace.png",width=2000,height=1000,res=200)
makeplot.topology(octo.trees, burnin = 50)$trace.plot
dev.off()
```
![Tree Topology trace](Topology_trace.png)
### Split frequency plots
Next up, we look at the split frequency plots. According to the documentation for the functions:
"In an analysis that is mixing well, we expect the frequencies in the sliding window to move around a lot, but the frequencies in the cumulative plot should level off. If the frequencies in the cumulative plot have not leveled off, it might be necessary to run the chain for longer." 

```{r RWTY sliding split frequencies, eval=F}
png("sliding_split_freq.png",width=2000,height=1000,res=200)
makeplot.splitfreqs.sliding(octo.trees, burnin = 50)
dev.off()
```

![Sliding Window Split Frequency Plot](sliding_split_freq.png)

```{r RWTY cumulative split frequencies, eval=F}
png("cumulative_split_freq.png",width=2000,height=1000,res=200)
makeplot.splitfreqs.cumulative(octo.trees, burnin = 50)
dev.off()
```
![Cumulative split frequency plot](cumulative_split_freq.png)

### Tree space heatmap
This plot show the areas of treespace sampled by each run.

```{r RWTY treespace plot, eval=F}
png("treespace_heatmap.png",width=2000,height=1000,res=200)
makeplot.treespace(octo.trees, burnin =50, fill.color = "LnL")$treespace.heatmap
dev.off()
```

![Tree space heatmap](treespace_heatmap.png)

### Split Frequency Matrix

```{r RWTY split frequency matrix plot, eval=F}
png("split_freq_matrix.png",width=2000,height=1000,res=200)
makeplot.splitfreq.matrix(octo.trees, burnin = 50)$splitfreq.matrix
dev.off()
```

![Split frequency matrix plot](split_freq_matrix.png)

Finally, I am removing the octo.trees and octo.rwty object because they are super big.
```{r, warning=F, message=F}
rm(octo.trees)
rm(octo.rwty)
```


# Single Gene Trees
In the final figure I want to also include trees constructed from each gene individually as well. This section of the code follows the same process as the multi-gene tree section, but with only one gene at a time.  

## 12s Only Tree
```{r}
x12s.solo=x12s.final[complete.cases(access$X12s),]
write.nexus.data(x12s.solo,"12s_solo.nxs",charsperline = 1000)
length(x12s.solo[1,])+max(nchar(labels(x12s.solo)))+10
```


```{bash}
sed -i -E 's/(^.{536}).*/\1/g' 12s_solo.nxs
```

### Writing length variable lines
```{r}
write.table(paste("  DIMENSIONS NTAX=",
                  dim(x12s.solo)[1],
                  " NCHAR=",
                  dim(x12s.solo)[2],
                  ";",sep=""),
            "12s_dimensions",
            row.names = F,col.names = F,quote = F)

write.table(models.muus[grepl("\\(1\\)|\\(2\\)",models.muus)],
            "12s_models_to_write",row.names = F,col.names = F,quote = F)
```


### Generating nexus file with MrBayes command block. 
```{bash, eval=F}
rm 12s_alone.nxs #just cleaning out previous iteration if running multiple times
cp -r mb_12s mb_12s.backup
rm -r mb_12s
touch 12s_alone.nxs
echo "#NEXUS" >> 12s_alone.nxs
echo "BEGIN DATA;" >> 12s_alone.nxs
sed -n 1p 12s_dimensions >> 12s_alone.nxs
echo "  FORMAT DATATYPE=DNA MISSING=? GAP=- INTERLEAVE=YES;
  MATRIX
  
  [12s]" >> 12s_alone.nxs 
sed -n 7,96p 12s_solo.nxs >> 12s_alone.nxs
echo "
begin mrbayes;
	[Define pairs for the doublet model]" >> 12s_alone.nxs
cat basepairs | sed '1s/^/    pairs /' | sed -E 's/(:[0-9]{,4})/\1,/g' | sed -E 's/,$/;/g' | sed -r 's/(.{72,76} )/\1\n         /g' >> 12s_alone.nxs
echo " " >> 12s_alone.nxs
cat stems12s | sed '1s/^/    	charset  12s-stems = /' | sed -E 's/$/;/g' | sed -r 's/(.{73,76} )/\1\n           /g' | sed '1s/^/\n/' >> 12s_alone.nxs

echo " " >> 12s_alone.nxs
cat loops12s | sed '1s/^/    	charset  12s-loops = /' | sed -E 's/$/;/g' | sed -r 's/(.{73,76} )/\1\n           /g' | sed '1s/^/\n/' >> 12s_alone.nxs
echo " " >> 12s_alone.nxs
echo "      partition by_gene = 2:12s-stems,12s-loops;" >> 12s_alone.nxs
echo "	    set partition = by_gene;
" >> 12s_alone.nxs
echo "	    constraint outg = Octopus_vulgaris Octopus_rubescens Octopus_bimaculoides Octopus_cyanea Amphioctopus_aegina Hapalochlaena_maculosa Abdopus_aculeatus Octopus_tetricus Octopus_berrima Macroctopus_maorum;
	    prset topologypr = constraints (outg);
	    outgroup Octopus_vulgaris;
	    " >> 12s_alone.nxs
echo "	    lset applyto=(1)   nucmodel=doublet;
	    lset applyto=(2)   nucmodel=4by4;
	    prset ratepr=variable;
	    unlink statefreq=(all) revmat=(all) shape=(all) pinvar=(all);" >> 12s_alone.nxs

cat 12s_models_to_write >> 12s_alone.nxs

echo "
      mcmcp ngen=100000000 printfreq=1000 samplefreq=10000 stoprule=yes stopval=0.01 nruns=5 nchains=2 burninfrac=0.25;
      mcmc;
      sump;
      sumt filename=mb_12s/12s_alone.nxs conformat=simple;

end;
" >> 12s_alone.nxs

mkdir mb_12s
cp 12s_alone.nxs mb_12s/12s_alone.nxs
```

### Running the MrBayes analysis

```{bash, eval=F}
mpiexec -np 10 mb mb_12s/12s_alone.nxs > mb_12s_log
```

## COIII Only Tree
```{r}
coiii.solo=coiii.final[complete.cases(access$COIII)]
write.nexus.data(coiii.solo,"coiii_solo.nxs",charsperline = 1000)
length(coiii.solo$Octopus_vulgaris)+max(nchar(labels(coiii.solo)))+10
```


```{bash}
sed -i -E 's/(^.{711}).*/\1/g' coiii_solo.nxs
```

### Writing length variable lines
```{r}
write.table(paste("  DIMENSIONS NTAX=",
                  length(coiii.solo),
                  " NCHAR=",
                  length(coiii.solo$Octopus_vulgaris),
                  ";",sep=""),
            "coiii_dimensions",
            row.names = F,col.names = F,quote = F)
coiii.models=models.muus[grepl("\\(3\\)|\\(4\\)|\\(5\\)",models.muus)]
coiii.models=gsub("\\(3\\)","(1)",coiii.models)
coiii.models=gsub("\\(4\\)","(2)",coiii.models)
coiii.models=gsub("\\(5\\)","(3)",coiii.models)
write.table(coiii.models,
            "coiii_models_to_write",row.names = F,col.names = F,quote = F)

write.table(data.frame(models=c(
            paste("      charset coiii_pos1 =  ",1," - ",length(coiii.solo$Octopus_vulgaris),"\\3;",sep=""),
            paste("      charset coiii_pos2 =  ",2," - ",length(coiii.solo$Octopus_vulgaris),"\\3;",sep=""),
            paste("      charset coiii_pos3 =  ",3," - ",length(coiii.solo$Octopus_vulgaris),"\\3;",sep="")
            )),
            "charset.coii2",
            row.names = F,col.names = F,quote = F)
```


### Generating nexus file with MrBayes command block. 
```{bash, eval=F}
rm coiii_alone.nxs #just cleaning out previous iteration if running multiple times
cp -r mb_coiii mb_coiii.backup
rm -r mb_coiii
touch coiii_alone.nxs
echo "#NEXUS" >> coiii_alone.nxs
echo "BEGIN DATA;" >> coiii_alone.nxs
sed -n 1p coiii_dimensions >> coiii_alone.nxs
echo "  FORMAT DATATYPE=DNA MISSING=? GAP=- INTERLEAVE=YES;
  MATRIX
  
  [COIII]" >> coiii_alone.nxs
sed -n 7,96p coiii_solo.nxs >> coiii_alone.nxs
echo "
begin mrbayes;" >> coiii_alone.nxs
echo " " >> coiii_alone.nxs
sed -n 1,3p charset.coii2 >> coiii_alone.nxs
echo " " >> coiii_alone.nxs
echo "      partition by_gene = 3:coiii_pos1,coiii_pos2,coiii_pos3;" >> coiii_alone.nxs
echo "	    set partition = by_gene;
" >> coiii_alone.nxs
echo "	    constraint outg = Octopus_vulgaris Octopus_rubescens Octopus_bimaculoides Octopus_cyanea Amphioctopus_aegina Hapalochlaena_maculosa Abdopus_aculeatus Octopus_tetricus Octopus_berrima Macroctopus_maorum;
	    prset topologypr = constraints (outg);
	    outgroup Octopus_vulgaris;
	    " >> coiii_alone.nxs
echo "	    lset applyto=(all)   nucmodel=4by4;
	    prset ratepr=variable;
	    unlink statefreq=(all) revmat=(all) shape=(all) pinvar=(all);" >> coiii_alone.nxs

cat coiii_models_to_write >> coiii_alone.nxs

echo "
      mcmcp ngen=100000000 printfreq=1000 samplefreq=10000 stoprule=yes stopval=0.01 nruns=5 nchains=2 burninfrac=0.25;
      mcmc;
      sump;
      sumt filename=mb_coiii/coiii_alone.nxs conformat=simple;

end;
" >> coiii_alone.nxs

mkdir mb_coiii
cp coiii_alone.nxs mb_coiii/coiii_alone.nxs
```


### Running the MrBayes analysis
```{bash, eval=F}
mpiexec -np 10 mb mb_coiii/coiii_alone.nxs > mb_coiii_log
```

## CytB Only Tree
```{r}
cytb.solo=cytb.final[complete.cases(access$Cytb)]
write.nexus.data(cytb.solo,"cytb_solo.nxs",charsperline = 1000)
length(cytb.solo$Octopus_vulgaris)+max(nchar(labels(cytb.solo)))+10
```


```{bash}
sed -i -E 's/(^.{675}).*/\1/g' cytb_solo.nxs
```

### Writing length variable lines
```{r}
write.table(paste("  DIMENSIONS NTAX=",
                  length(cytb.solo),
                  " NCHAR=",
                  length(cytb.solo$Octopus_vulgaris),
                  ";",sep=""),
            "cytb_dimensions",
            row.names = F,col.names = F,quote = F)
cytb.models=models.muus[grepl("\\(6\\)|\\(7\\)|\\(8\\)",models.muus)]
cytb.models=gsub("\\(6\\)","(1)",cytb.models)
cytb.models=gsub("\\(7\\)","(2)",cytb.models)
cytb.models=gsub("\\(8\\)","(3)",cytb.models)
write.table(cytb.models,
            "cytb_models_to_write",row.names = F,col.names = F,quote = F)

write.table(data.frame(models=c(
            paste("      charset cytb_pos1 =  ",1," - ",length(cytb.solo$Octopus_vulgaris),"\\3;",sep=""),
            paste("      charset cytb_pos2 =  ",2," - ",length(cytb.solo$Octopus_vulgaris),"\\3;",sep=""),
            paste("      charset cytb_pos3 =  ",3," - ",length(cytb.solo$Octopus_vulgaris),"\\3;",sep="")
            )),
            "charset.cytb2",
            row.names = F,col.names = F,quote = F)
```


### Generating nexus file with MrBayes command block. 
```{bash, eval=F}
rm cytb_alone.nxs #just cleaning out previous iteration if running multiple times
cp -r mb_cytb mb_cytb.backup
rm -r mb_cytb
touch cytb_alone.nxs
echo "#NEXUS" >> cytb_alone.nxs
echo "BEGIN DATA;" >> cytb_alone.nxs
sed -n 1p cytb_dimensions >> cytb_alone.nxs
echo "  FORMAT DATATYPE=DNA MISSING=? GAP=- INTERLEAVE=YES;
  MATRIX
  
  [Cytb]" >> cytb_alone.nxs
sed -n 7,98p cytb_solo.nxs >> cytb_alone.nxs
echo "
begin mrbayes;" >> cytb_alone.nxs
echo " " >> cytb_alone.nxs
sed -n 1,3p charset.cytb2 >> cytb_alone.nxs
echo " " >> cytb_alone.nxs
echo "      partition by_gene = 3:cytb_pos1,cytb_pos2,cytb_pos3;" >> cytb_alone.nxs
echo "	    set partition = by_gene;
" >> cytb_alone.nxs
echo "	    constraint outg = Octopus_vulgaris Octopus_bimaculoides Octopus_cyanea Amphioctopus_aegina Hapalochlaena_maculosa Abdopus_aculeatus Octopus_tetricus Octopus_berrima Macroctopus_maorum;
	    prset topologypr = constraints (outg);
	    outgroup Octopus_vulgaris;
	    " >> cytb_alone.nxs
echo "	    lset applyto=(all)   nucmodel=4by4;
	    prset ratepr=variable;
	    unlink statefreq=(all) revmat=(all) shape=(all) pinvar=(all);" >> cytb_alone.nxs

cat cytb_models_to_write >> cytb_alone.nxs

echo "
      mcmcp ngen=100000000 printfreq=1000 samplefreq=10000 stoprule=yes stopval=0.01 nruns=5 nchains=2 burninfrac=0.25;
      mcmc;
      sump;
      sumt filename=mb_cytb/cytb_alone.nxs conformat=simple;

end;
" >> cytb_alone.nxs

mkdir mb_cytb
cp cytb_alone.nxs mb_cytb/cytb_alone.nxs
```
### Running the MrBayes analysis
```{bash, eval=F}
mpiexec -np 10 mb mb_cytb/cytb_alone.nxs > mb_cytb_log
```


# Reading into R the resulting trees

## Reading in the multi-gene tree
The file that contains the consensus trees is "complete.nxs.con.tre". This file actually contains two trees.  One contains no branch probability information, and the other contains the branch probabilities.  The following code deletes the tree without probabilities from file. If both trees are present, R opens the tree as a multi-phylo class, and manipulation of the tree becomes much harder.
```{bash eval=FALSE}
cp ./mb/complete.nxs.con.tre ./mb/completeBACKUP.nxs.con.tre
sed -i '/Note: This tree contains information only on the topology/d' ./mb/complete.nxs.con.tre
sed -i '/and branch lengths (median of the posterior probability density)/d' ./mb/complete.nxs.con.tre
sed -i '/):/d' ./mb/complete.nxs.con.tre
```


Now I read the tree into R using the following code.
```{r}
mrbayes.tree=read.nexus("mb/complete.nxs.con.tre",tree.names=NULL)
```

## Reading the 12s tree
Now the same or the 12s tree
```{bash eval=FALSE}
cp ./mb_12s/12s_alone.nxs.con.tre ./mb_12s/12s_aloneBACKUP.nxs.con.tre
sed -i '/Note: This tree contains information only on the topology/d' ./mb_12s/12s_alone.nxs.con.tre
sed -i '/and branch lengths (median of the posterior probability density)/d' ./mb_12s/12s_alone.nxs.con.tre
sed -i '/):/d' ./mb_12s/12s_alone.nxs.con.tre
```


Now I read the tree into R using the following code.
```{r}
x12s.tree=read.nexus("mb_12s/12s_alone.nxs.con.tre",tree.names=NULL)
```


## Reading the COIII tree
Now the same or the coiii tree
```{bash eval=FALSE}
cp ./mb_coiii/coiii_alone.nxs.con.tre ./mb_coiii/coiii_aloneBACKUP.nxs.con.tre
sed -i '/Note: This tree contains information only on the topology/d' ./mb_coiii/coiii_alone.nxs.con.tre
sed -i '/and branch lengths (median of the posterior probability density)/d' ./mb_coiii/coiii_alone.nxs.con.tre
sed -i '/):/d' ./mb_coiii/coiii_alone.nxs.con.tre
```


Now I read the tree into R using the following code.
```{r}
coiii.tree=read.nexus("mb_coiii/coiii_alone.nxs.con.tre",tree.names=NULL)
```


## Reading the CytB tree
Now the same for the cytb tree
```{bash eval=FALSE}
cp ./mb_cytb/cytb_alone.nxs.con.tre ./mb_cytb/cytb_aloneBACKUP.nxs.con.tre
sed -i '/Note: This tree contains information only on the topology/d' ./mb_cytb/cytb_alone.nxs.con.tre
sed -i '/and branch lengths (median of the posterior probability density)/d' ./mb_cytb/cytb_alone.nxs.con.tre
sed -i '/):/d' ./mb_cytb/cytb_alone.nxs.con.tre
```


Now I read the tree into R using the following code.
```{r}
cytb.tree=read.nexus("mb_cytb/cytb_alone.nxs.con.tre",tree.names=NULL)
```



# Plotting the tree
Finally, I get to plot the trees produced by this analysis.  

```{r}
tips.add=mrbayes.tree$tip.label
tips.add[!grepl("leioderma",tips.add)]=""
prefix="                                            "

tips.add[tips.add=="Muusoctopus_leioderma_neotype"]=paste(prefix,"(neotype)")
#tips.add[tips.add=="Muusoctopus_leioderma_burrows"]=paste(prefix,"(Burrows Bay)")
tips.add[tips.add=="Muusoctopus_leioderma_casiz31213"]=paste(prefix,"(CASIZ 31213)")

mrbayes.tree$tip.label[mrbayes.tree$tip.label=="Muusoctopus_leioderma_neotype"]="Muusoctopus leioderma"
#mrbayes.tree$tip.label[mrbayes.tree$tip.label=="Muusoctopus_leioderma_burrows"]="Muusoctopus leioderma"
mrbayes.tree$tip.label[mrbayes.tree$tip.label=="Muusoctopus_leioderma_casiz31213"]="Muusoctopus leioderma"
```


Making vectors for each tree that contain color coding by clade.
```{r}
colors.tip=rep("black",length(mrbayes.tree$tip.label))
colors.tip[grep("Muusoctopus",mrbayes.tree$tip.label)]="red" #"#e06565ff"#"coral3"
colors.tip[grep("leioderma|Burrows",mrbayes.tree$tip.label)]="blue" #"#2f2c86ff"#"darkcyan"
colors.tip[grep("Enteroctopus|Sasakiopus|californicus",mrbayes.tree$tip.label)]="darkgreen" #"#289e31ff"#"darkolivegreen"


cytb.colors.tip=rep("black",length(cytb.tree$tip.label))
cytb.colors.tip[grep("Muusoctopus",cytb.tree$tip.label)]="red" #"#e06565ff"#"coral3"
cytb.colors.tip[grep("leioderma|Burrows",cytb.tree$tip.label)]="blue" #"#2f2c86ff"#"darkcyan"
cytb.colors.tip[grep("Enteroctopus|Sasakiopus|californicus",cytb.tree$tip.label)]="darkgreen" #"#289e31ff"#"darkolivegreen"

coiii.colors.tip=rep("black",length(coiii.tree$tip.label))
coiii.colors.tip[grep("Muusoctopus",coiii.tree$tip.label)]="red" #"#e06565ff"#"coral3"
coiii.colors.tip[grep("leioderma|Burrows",coiii.tree$tip.label)]="blue" #"#2f2c86ff"#"darkcyan"
coiii.colors.tip[grep("Enteroctopus|Sasakiopus|californicus",coiii.tree$tip.label)]="darkgreen" #"#289e31ff"#"darkolivegreen"

x12s.colors.tip=rep("black",length(x12s.tree$tip.label))
x12s.colors.tip[grep("Muusoctopus",x12s.tree$tip.label)]="red" #"#e06565ff"#"coral3"
x12s.colors.tip[grep("leioderma|Burrows",x12s.tree$tip.label)]="blue" #"#2f2c86ff"#"darkcyan"
x12s.colors.tip[grep("Enteroctopus|Sasakiopus|californicus",x12s.tree$tip.label)]="darkgreen" #"#289e31ff"#"darkolivegreen"

```


```{r}
cytb.tree$tip.label=
  gsub("_casiz31213|_neotype|_akambei|_longibrachus$","",cytb.tree$tip.label)
cytb.tree$tip.label=
  gsub("(\\w)\\w+_(\\w+)","\\1._\\2",cytb.tree$tip.label)
cytb.tree$tip.label=
  gsub("B._octopus","Burrows_Bay",cytb.tree$tip.label)


coiii.tree$tip.label=
  gsub("_casiz31213|_neotype|_akambei|_longibrachus$","",coiii.tree$tip.label)
coiii.tree$tip.label=
  gsub("(\\w)\\w+_(\\w+)","\\1._\\2",coiii.tree$tip.label)
coiii.tree$tip.label=
  gsub("B._octopus","Burrows_Bay",coiii.tree$tip.label)

x12s.tree$tip.label=
  gsub("_casiz31213|_neotype|_akambei|_longibrachus$","",x12s.tree$tip.label)
x12s.tree$tip.label=
  gsub("(\\w)\\w+_(\\w+)","\\1._\\2",x12s.tree$tip.label)
x12s.tree$tip.label=
  gsub("B._octopus","Burrows_Bay",x12s.tree$tip.label)


```


```{r}
nodes95=rep(NA,length(mrbayes.tree$node.label))
nodes100=rep(NA,length(mrbayes.tree$node.label))
#nodes[mrbayes.tree$node.label>0]="♦"
#nodes[mrbayes.tree$node.label>0.95]="■"
nodes95[mrbayes.tree$node.label>0.95&mrbayes.tree$node.label<0.99999999]="┼"
nodes100[mrbayes.tree$node.label>0.9999999]="●"


#ramp=blue2green2red(1001)
ramp=hcl.colors(1001,palette="viridis")
node.colors=rep("black",length(mrbayes.tree$node.label))
for (i in 2:length(node.colors)){
  node.colors[i]=ramp[as.numeric(mrbayes.tree$node.label[i])*1000]
}

#nodes[mrbayes.tree$node.label==1]="†"
#nodes[mrbayes.tree$node.label>=0.95&mrbayes.tree$node.label<1]="‡"
#nodes[mrbayes.tree$node.label>=0.90&mrbayes.tree$node.label<0.95]="◊"
```



```{r, }
svg(file="Figure2.svg",width=8,height=12)
  par(fig=c(0,1,0.31,1))
  par(mai=c(0,0,0,0))
  
  plot(mrbayes.tree,show.node.label=F,use.edge.length=T,align.tip.label=F,edge.width=2,
     show.tip.label = T,x.lim=1.65,font=3,tip.color = colors.tip)
  nodelabels(c(NA,rep("♦",length(mrbayes.tree$node.label)-1)), adj = c(0.5, 0.5), frame = "n",
             cex = 1,font=2,col=node.colors)
  nodelabels(c(NA,rep("◊",length(mrbayes.tree$node.label)-1)), adj = c(0.45, 0.53), frame = "n",
             cex = 1.3,font=2,col="black")
  nodelabels(nodes95, adj = c(0.46, 0.43), frame = "n",
             cex = 0.5,font=2,col="black")
  nodelabels(nodes100, adj = c(0.4, 0.43), frame = "n",
             cex = 0.5,font=2,col="black")
  tiplabels(tips.add,frame="n",adj=c(0,0.3),col="blue") #"#2f2c86ff")
#  nodelabels(mrbayes.tree$node.label,bg=NULL,cex=0.3,frame="none",adj=c(1.1,-0.1))
    
    
# Posterior Probability key
  key.start=1.15
  key.end=1.65
  key.top=26
  key.bottom=25.5
  interp=round(seq(from=0,to=1,by=0.001),3)

  key.points=seq(from=key.start,to=key.end,length.out=length(interp))
  for(i in 1:length(key.points)){
    lines(rep(key.points[i],2),c(key.bottom,key.top),col=ramp[i],lwd=2)
  }
  lines(c(key.start,key.end,key.end,key.start,key.start),
        c(key.bottom,key.bottom,key.top,key.top,key.bottom),lwd=2)
  text(key.points[c(1,length(interp)/2,length(interp))],rep(key.top+0.5,3)
     ,c(0,0.5,1),cex=1,font=1)
  lines(c(key.points[length(interp)/2],
          key.points[length(interp)/2]),
        c(key.top+0.2,key.top),lwd=2)
  lines(c(key.points[1],key.points[1]),c(key.top+0.2,key.top),lwd=2)
  lines(c(key.points[length(interp)],key.points[length(interp)]),c(key.top+0.2,key.top),lwd=2)
  text((key.start+key.end)/2,key.top+1.1,'posterior probability')

  points(key.end,key.bottom-0.5,pch="♦",col=ramp[999])
  points(key.end+0.0002,key.bottom-0.525,pch="◊",cex=1.3)
  points(key.end+0.0005,key.bottom-0.55,pch="┼",cex=0.5,font=2)
  text(key.end,key.bottom-0.5,"0.95<pp<1.00:  ",adj=c(1,0.5))

  points(key.end,24,pch="♦",col=ramp[1001])
  points(key.end+0.0002,23.975,pch="◊",cex=1.3)
  points(key.end+0.0006,23.98,pch="●",cex=0.5,font=2)
  text(key.end,24,"pp=1.00:  ",adj=c(1,0.5))

# Lines denoting clades of interest
  lines(c(1.5,1.5),c(10,20),col="red",lwd=5)
  text(1.54,15,"genus Muusoctopus",srt=-90,col="red",cex=1.5)

  lines(c(1.6,1.6),c(6,22),col="darkgreen",lwd=5)
  text(1.64,15,"family Enteroctopodidae",srt=-90,col="darkgreen",cex=1.5)

#  lines(c(0,1),c(0,50))


  par(fig=c(0,0.33,0,0.3),new=T)

  plot(cytb.tree,show.node.label=F,use.edge.length=T,align.tip.label=F,edge.width=1.5,
     show.tip.label = T,x.lim=0.8,font=3,tip.color = cytb.colors.tip,cex=0.7)
  mtext("CytB",cex=1.5)

  par(fig=c(0.33,0.66,0,0.3),new=T)

  plot(coiii.tree,show.node.label=F,use.edge.length=T,align.tip.label=F,edge.width=1.5,
     show.tip.label = T,x.lim=1.8,font=3,tip.color = coiii.colors.tip,cex=0.7)
  mtext("COIII",cex=1.5)

  par(fig=c(0.66,1,0,0.3),new=T)

  plot(x12s.tree,show.node.label=F,use.edge.length=T,align.tip.label=F,edge.width=1.5,
     show.tip.label = T,x.lim=0.9,font=3,tip.color = x12s.colors.tip,cex=0.7)
  mtext("12s",cex=1.5)

dev.off()
```

This is bash code to convert the svg to png that can be displayed in the RMarkdown pdf.
```{bash}
cairosvg Figure2.svg -o Figure2.png -d 300
```

And this is bash code to convert the svg to eps for publication.
```{bash}
inkscape Figure2.svg -o Figure2.eps --export-ignore-filters --export-ps-level=3
```


![Phylogeny of family Enteroctopodidae](Figure2.png)